File: concepts/03_errors_and_resources/03_error_chaining_suppression_and_aggregation/cpp/test_02_multiple_cleanup_failures.cpp
// 多个清理失败的保留方式。
// 共同问题：主体失败后多个清理也失败时，哪个错误成为最外层；其余失败如何保留；
// 没有内置聚合协议时如何显式保存全部结果。
//
// polyglot-family: errors_and_resources
// polyglot-concept: error_chaining_suppression_and_aggregation
// polyglot-related: languages/cpp/language/test_012_exceptions_raii_and_noexcept.cpp

#include <gtest/gtest.h>

#include <exception>
#include <functional>
#include <stdexcept>
#include <string>
#include <vector>

namespace {

std::vector<std::exception_ptr> close_all(
    const std::vector<std::function<void()>>& cleanups) {
  std::vector<std::exception_ptr> failures;
  for (auto cleanup = cleanups.rbegin(); cleanup != cleanups.rend(); ++cleanup) {
    try {
      (*cleanup)();
    } catch (...) {
      failures.push_back(std::current_exception());
    }
  }
  return failures;
}

std::string message_from(const std::exception_ptr& failure) {
  try {
    std::rethrow_exception(failure);
  } catch (const std::exception& error) {
    return error.what();
  } catch (...) {
    return "<non-standard exception>";
  }
}

TEST(MultipleFailureConcept, ExplicitCleanupRunnerCollectsEveryFailureInLifoOrder) {
  std::vector<std::string> events;
  const std::vector<std::function<void()>> cleanups{
      [&] {
        events.push_back("cleanup:first");
        throw std::runtime_error{"first"};
      },
      [&] {
        events.push_back("cleanup:second");
        throw std::logic_error{"second"};
      },
  };

  const std::vector<std::exception_ptr> failures = close_all(cleanups);

  EXPECT_EQ(
      events,
      (std::vector<std::string>{"cleanup:second", "cleanup:first"}));
  ASSERT_EQ(failures.size(), 2U);
  EXPECT_EQ(message_from(failures[0]), "second");
  EXPECT_EQ(message_from(failures[1]), "first");

  // C++20 没有自动 SuppressedError/ExceptionGroup。noexcept 析构负责安全收尾；需要报告
  // 多个清理错误时，应在非析构控制器中显式捕获 exception_ptr，避免双异常 terminate。
}

}  // namespace

File: concepts/03_errors_and_resources/03_error_chaining_suppression_and_aggregation/julia/test_02_multiple_cleanup_failures.jl
# polyglot-family: errors_and_resources
# polyglot-concept: error_chaining_suppression_and_aggregation
# polyglot-related: languages/julia/tooling_and_runtime/12_tasks_channels_and_cooperative_concurrency/
# polyglot-related+: test_093_task_failure_and_composite_exception.jl
#
# 共同问题：并发或多个 cleanup 同时失败时是否丢失错误。
# 对照观察：@sync 等待其词法子任务，并以 CompositeException 聚合多个 TaskFailedException。

using Test

@testset "结构化并发聚合多个失败" begin
    exception = try
        @sync begin
            @async error("first")
            @async throw(ArgumentError("second"))
        end
        nothing
    catch caught
        caught
    end
    @test exception isa CompositeException
    @test length(exception.exceptions) == 2
    @test all(item -> item isa TaskFailedException, exception.exceptions)
end
